### ISSUE TITLE
Bring Cyclic to parity with Block

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#17764]
Our `Block` distribution has received a lot of attention over the years in terms of capabilities and optimizations; by contrast, the `Cyclic` distribution has not received nearly as much attention.  However, `Cyclic` can sometimes be more powerful in terms of load balancing or growing arrays over time, and it has the advantage of being nearly as simple as `Block`.

This issue proposes that we should bring the two into parity, in terms of:

* features: anything a user can do on `Block` they should also be able to do on `Cyclic`, within reason
  * e.g., iteration, slicing, queries — I think these are mostly close to parity, but don't feel confident
  * sparse subdomains: I'm fairly certain that this is not supported for `Cyclic` yet
* performance: generally, `Cyclic` should behave as similarly to `Block` as possible for idioms that are independent of distribution, such as:
  * embarrassingly parallel computations (e.g., stream variants)
  * random access (local vs. remote)

**Comment by ronawho** [comment]
On the performance side, here's the differences I know of:
 - We don't optimize indexing as heavily so indexing with limited comm is slower (`studies/bale/aggregation/ig.chpl`)
 - We don't implement fast-followers so things like stream are slower (`release/examples/benchmarks/hpcc/stream.chpl`)
 - We haven't optimized scans and I'm  not sure if it's possible to achieve parity for scans (`scan/scanPerf.chpl`)
 - I don't think we've optimized bulk comm so large transfers will be slower (not sure what benchmarks we have offhand)

**Comment by bradcray** [comment]
Good point on scans.  As best I know, in order to do parallel scans on cyclic, the communication volumes will be similar in size to a given locale's local chunk (so, `n/numLocales`, say).

**Comment by mppf** [comment]
It looks to me like bulk transfer for Cyclic is disabled. Several important functions have `where useBulkTransferDist` which has this comment:

```
  // Toggles the functionality to perform strided bulk transfers involving
  // distributed arrays.
  //
  // Currently disabled due to observations of higher communication counts
  // compared to element-by-element assignment.
  @chpldoc.nodoc
  config param useBulkTransferDist = false;
```

I guess we view Cyclic array bulk transfer as a strided operation? But MyCyclicArray = MyOtherCyclicArray isn't a strided transfer, if they have the same domain.
